### PR TITLE
New `RedisExampleDatabase`, `MultiplexedDatabase`, and `ReadOnlyDatabase` for team workflows

### PR DESCRIPTION
**Pull request by Zac-HD**
Our `ExampleDatabase` is a lovely thing - and persisting failing examples is great - but our story for *sharing* them is historically not amazing.  "Sync files or put them in version control" does *work*, but gets painful to set up or very noisy, respectively.  This PR is because after sketching out my ideal fuzzing workflow for teams, I realised it was just as useful for CI .  In short:

- The CI server (and/or fuzz server) saves any failing examples to a *network* datastore.
- Developer machines read from the shared datastore, maintain a local copy, and may or may not write to it.
- To reproduce failing examples locally, *just run the tests*.

And while of course this is optional and doesn't change the default directory-based DB, if you have a Redis server it's now ~ten lines of setup code to get the workflow above.  [Here's the new docs](https://hypothesis--2622.org.readthedocs.build/en/2622/database.html).

### CODE REVIEW

File: hypothesis-python/docs/database.rst
Comment (DRMacIver) [inline hypothesis-python/docs/database.rst:25]: I feel like removing this entirely as a top level piece of documentation makes it harder to understand Hypothesis's default behaviour. I'm not quite sure what needs to go here in its place, but I think something to the tune of "By default Hypothesis uses a `DirectoryBasedExampleDatabase` (linked) which etc" here would be helpful.

**Review by DRMacIver — COMMENTED** [review]
One comment about documentation but I don't need to rereview after. Sorry for the long delay on getting around to reviewing this!

File: hypothesis-python/docs/database.rst
Comment (Zac-HD) [reply hypothesis-python/docs/database.rst:25]: That's just moved [one paragraph down](https://hypothesis--2622.org.readthedocs.build/en/2622/database.html#exampledatabase-implementations) to the first sentence under "Implementations".  Happy to move that if you have a preferred location?

(also can't merge until you hit `approve`, so you'll need to do at least the null review)

File: hypothesis-python/src/hypothesis/database.py
Comment (kxepal) [inline hypothesis-python/src/hypothesis/database.py:252]: For readonly database I would expect here some ReadOnly / NotImplementedError exception, rather than silent ok on write modifications. I guess, this implementation was with a reason, but may be notice that behaviour in docs? Or may be name JustForReadDatabase would be a bit clear? Read-only thing are quite nervous about write operations in common...

File: hypothesis-python/src/hypothesis/extra/redis.py
Comment (kxepal) [inline hypothesis-python/src/hypothesis/extra/redis.py:35]: I guess prefix would be better to be configured via some init argument. Imagine single redis..well, hypothesis server which maintains multiple teams. Some of them would like to run GC on their examples for the fresh start. That would be a problem to not touch others examples. And that single redis could be actually a common cluster.

File: hypothesis-python/src/hypothesis/extra/redis.py
Comment (kxepal) [inline hypothesis-python/src/hypothesis/extra/redis.py:45]: Any reasons for \n here? This may looks pretty, but repr isn't about pretty things. Also this like is good fit the single one with sane limits. For MultiplexedDatabase there is no such pretty left and that is ok.

File: hypothesis-python/src/hypothesis/extra/redis.py
Comment (kxepal) [inline hypothesis-python/src/hypothesis/extra/redis.py:37]: Why 8? Just curious...

Comment (kxepal) [reply hypothesis-python/src/hypothesis/extra/redis.py:37]: Also, why expiration logic is backend specific? Shouldn't it be more general and backend would just implement required routines to maintain it?

**Review by kxepal — COMMENTED** [review]
In general LGFM, left a few questions about implementation. This reminds me about CouchDB backend, but nowdays there are no usecases for me to it /: On other hand redis as good as simple to keep it running for this goal.

File: hypothesis-python/src/hypothesis/database.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/database.py:252]: Yeah, the idea is that you can use it in contexts where you want write/delete operations to be noops but not raise errors.  I'll explicitly document that though, thanks!

File: hypothesis-python/src/hypothesis/extra/redis.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/extra/redis.py:37]: Eight because I was thinking about weekly cron jobs - and you don't want expire keys a few seconds before they're accessed.

Expiration logic is Redis-specific because it's more difficult to clear out Redis keys by age than delete local directories, and because unlike the others there's a non-Hypothesis server responsible for expiration.

Comment (kxepal) [reply hypothesis-python/src/hypothesis/extra/redis.py:37]: After some thoughts, I agree with that.  How about to make this behaviour more screaming? Like, with `..note` / `..warning` sphinx things? I was trying to write about "let's document this" at first place, but after third iteration of reading code I noticed "weekly" word - suddenly, for nonnative speaker this may not be a stop word to realize some important behaviour. Personal experience, but else I don't have. 

In the end, this behavior makes this backend different from others by ability to remove old data. This may be not always good. Imagine you went to Hawaii on vacation for two weeks and after return all your examples were vanished while no other developer ran any test (holidays!).

Pardon me, may be I just too strict about the thing that would delete some data by default without explicit asking about. May be another way would be to not expire anything at all, but let users to decide whenever they want do this and how? Much like with the other backends.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/extra/redis.py:37]: Note that it's not *old* data, it's *unaccessed* data - and cross-platform `atime` is a mess.

I'll make the expiry note much more obvious though.

Comment (kxepal) [reply hypothesis-python/src/hypothesis/extra/redis.py]: That would be a good clarification and it removes all my doubts. Thanks!

**Comment by Zac-HD** [comment]
@DRMacIver @kxepal @sobolevn an approving review would get this merged :wink:

**Review by kxepal — APPROVED** [review]
LGFM, now without any questions (: This feature really worth to try in prod. If anything was missed it could be fixed later.

**Comment by Zac-HD** [comment]
Cheers, I'm looking forward to it too!  I'll merge tomorrow, to give David time to respond to the docs point if he wants to :grin:

File: hypothesis-python/docs/database.rst
Comment (DRMacIver) [reply hypothesis-python/docs/database.rst:25]: So it is. Sorry, apparently I had reading comprehension fail. Glad @kxepal did a more thorough review than me and approved it!

**Comment by sobolevn** [comment]
For some reason this release breaks our CI: https://github.com/dry-python/returns/pull/656

**Comment by Zac-HD** [comment]
Sure it's not the entrrypoint patch? https://hypothesis.readthedocs.io/en/latest/changes.html#v5-36-2

**Pull request merged**
